plugins/codemirror/plugin.js

```diff
@@ -0,0 +1,130 @@
+(function() {
+	'use strict';
+
+	var stylesLoaded = false;
+
+
+	CKEDITOR.plugins.add('codemirror', {
+		_createCodeMirrorEditor: function(editor) {
+			var instance = this;
+
+			var codemirror = CKEDITOR.plugins.codemirror;
+
+			editor.addMode('source', function(callback) {
+				var contentsSpace = editor.ui.space('contents');
+
+				var textarea = contentsSpace.getDocument().createElement('textarea');
+				contentsSpace.append(textarea);
+
+				instance.codeMirrorEditor = CodeMirror.fromTextArea(
+					textarea.$,
+					{
+						lineNumbers: true,
+						mode: 'text/html'
+					}
+				);
+
+				var oldData = editor.getData(1);
+
+				instance.codeMirrorEditor.setValue(html_beautify(oldData));
+
+				instance.codeMirrorEditor.on('change', instance._handleCodeMirrorChange.bind(instance, editor, oldData));
+
+				editor.on('resize', instance._handleEditorResize.bind(instance));
+
+				editor.fire('ariaWidget', this);
+
+				callback();
+			});
+
+			editor.addCommand('codemirror', codemirror.commands.source);
+		},
+
+		_handleCodeMirrorChange: function(editor, oldData) {
+			var newData = this.codeMirrorEditor.getValue();
+
+			if (newData !== oldData) {
+				editor.setData(newData);
+			}
+		},
+
+		_handleEditorResize: function(event) {
+			this.codeMirrorEditor.setSize(null, event.data.outerHeight);
+		},
+
+		hidpi: true,
+
+		icons: 'source,source-rtl',
+
+		init: function(editor) {
+			var instance = this;
+
+			if (editor.plugins.detectConflict('codemirror', ['sourcearea'])) {
+				return;
+			}
+
+			if (editor.elementMode === CKEDITOR.ELEMENT_MODE_INLINE) {
+				return;
+			}
+
+			if (!stylesLoaded) {
+				CKEDITOR.document.appendStyleSheet(
+					this.path + 'deps/codemirror/codemirror.css'
+				);
+				stylesLoaded = true;
+			}
+
+			var instance = this;
+
+			CKEDITOR.scriptLoader.load([
+					instance.path + 'deps/codemirror/codemirror.js',
+					instance.path + 'deps/js-beautify/beautify-html.js'
+				],
+				function() {
+					CKEDITOR.scriptLoader.load(
+						instance.path + 'deps/codemirror/xml.js',
```

Comment:
I'm assuming that adding `xml.js` right after `codemirror.js` will fail because it might run before it?

This is just a nit, but it feels oddly specific and makes this harder to abstract... one option would be to concat this with `codemirror.js` during build time, since that's what we always want and that way we'd also be avoiding additional requests.

Reply:
> I'm assuming that adding xml.js right after codemirror.js will fail because it might run before it?

Yes exactly, `xml.js` [assumes](https://github.com/codemirror/CodeMirror/blob/6fcc49d5321261b525e50f111f3b28a602f01f71/mode/xml/xml.js#L5-L10) there's a global `CodeMirror` object or a `require` function to be able to load it (not necessarily in that order)

During development, I played around with this and managed to use `ugligyjs` to concatenate and minify both files into one, so I also think it would be possible to concatenate (just with `cat` or similar since CKBuilder takes care of minifying) both files as part of the build and it should keep on working.

Reply:
Could even cat `beautify` as well and have a single `vendor` file or something like that...

Anyways, this can be a small correction to apply once everything else is done.

Reply:
That's actually a good one, we can create a single deps file concatenating all of them during the build process so we can load only one here.